JSON serialization error handling

alfwatt opened the pull request:
Needs tests

Commit:
JSON serialization error handling

rclee reviewed (dismissed):
Thanks and congrats on PR #100 🎉

rclee commented:
Maybe we should add tests before we merge though

rclee committed:
TrustKit fix (#74)

* Add fix for Trustkit

Pulls in changes from TrustKit PR. Also updates project to match Xcode 10 settings

ref: https://github.com/mapbox/mapbox-events-ios/issues/66

* Call URLSession:didReceiveChallenge: completion block on the main queue

This is to fix an error on TrustKit logger tests

* Ensure handleChallenge: is called on a background queue

* Expose and call didReceiveChallenge: directly

Adds cedar tests for didRecieveChallenge: to confirm we are returning completion on the main thread when appropriate.

rclee committed:
Fix minor warnings (#102)

Fixes https://github.com/mapbox/mapbox-events-ios/issues/91 but we still need a better solution for keeping TrustKit up-to-date.

rclee committed:
Fix tests (#101)

* Fix tests

We need to replace `didRecieveChallenge` related test since we moved directions. Comparing the received disposition from the handler is more valuable. 

ref: https://github.com/mapbox/mapbox-events-ios/pull/74

* Cleanup tests

rclee committed:
Update version to 0.9.0 (#103)

Commit:
Debug Event Factory methods for Errors and Exceptions

Commit:
Expose pushEvent: to MMEAPIClient, use the MMEEevent constructors

Commit:
Whitespace and marks

Commit:
Expose `pushEvent:` to the test code
Add test for pushing error and exception events not ending up in the outbound queue (local logging only)

alfwatt committed:
Merge branch 'master' into issue-mt-252

alfwatt commented:
Added some simple tests to make sure Error and Exception events are only logged locally, not queued for upload

akitchen commented on MapboxMobileEventsTests/MMEEventsManagerTests.mm:
```diff
@@ -962,7 +964,36 @@ @interface MMEEventsManager (Tests) <MMELocationManagerDelegate>
+                NSLocalizedFailureReasonErrorKey: @"Test Error Failure Reason"
+            }];
+
+            [[MMEEventsManager sharedManager] pushEvent:[MMEEvent debugEventWithError:testError]];
```
You’ll want to put this inside a `beforeEach`, `it`, or `subjectAction` block for a more expected execution order. Otherwise it will be run when the spec is loaded, which might be too early. Same thing below.

akitchen reviewed (commented):
It’s fine to declare value objects for setup or expected results in a `context` or `describe` block (which are the same thing BTW) but the action portion of your tests should probably execute later, via a `beforeEach`, `it`, or `subjectAction` block.

Commit:
Move test actions into `'it` blocks

Commit:
Add testing section to readme.md
Gather all the Build and CI artifacts into a Packaging group

akitchen commented:
LGTM, especially from the testing perspective. I'll leave final approval up to the team 👍

rclee reviewed (approved):
Small typo -- everything else looks solid. Thanks Alf! 👍

julianrex commented on MapboxMobileEvents/MMEAPIClient.m:
```diff
@@ -66,21 +77,23 @@ - (void)postEvents:(NSArray *)events completionHandler:(nullable void (^)(NSErro
+                }
+                if (completionHandler) {
+                    error = error ?: statusError;
+                    completionHandler(error);
```
`completionHandler` will get called for each batch - is that correct?

Seems like it should only be called once? If so we might want to consider a `dispatch_group` here.

rclee replied:
Thanks for looking into this @julianrex! I'll try to give some more context behind these decisions. 

We wanted to prepare for deferred telemetry by implementing batching. Each batch represents the estimated event limit (around 1000) when sending a request. This should allow for us to send as many events as we desire from local storage and let `MMEAPIClient` handle the sending of those events. So, it seemed to me that each batch should be a separate call and return separately. Let me know if that makes sense.

julianrex replied:
Thanks for providing context, and apologies for leaping in without fully knowing the background 😉

Sounds like you need a callback per batch - a progress update I suppose?

Calling a completion block multiple times is non-standard and confusing (given that it's call "completion"), so perhaps we need two things here:
- a progress block (i.e. what you have so far, though passing in the block or some kind of block id might be useful)
- a completion block for when all the blocks have been posted.

julianrex commented on MapboxMobileEvents/MMEAPIClient.m:
```diff
@@ -66,21 +77,23 @@ - (void)postEvents:(NSArray *)events completionHandler:(nullable void (^)(NSErro
+                    [self.metricsManager updateMetricsFromEvents:events request:request error:error];
+                }
+            }];
+        }
```
If request is nil (and the completion block is called for each batch) should it be called here too?

julianrex replied:
np!

julianrex reviewed (commented):
Not really a review, just something I spotted that looked a bit off (also in the original code). I may be missing context, so apologies if I'm mistaken here!
